### PR TITLE
Add coding style guideline

### PR DESCRIPTION
**Pull request by elkoniu**
This PR contains updated coding guidelines.
It is based on previous documentation:
https://sssd.io/docs/developers/coding_style.html
https://www.port389.org/docs/389ds/development/coding-style.html

### CODE REVIEW

File: src/contrib/coding-style.rst [hunk 1]
```diff
@@ -1,12 +1,494 @@
...
+    /* Program logic here */
+
+    /*
+     * variable len - untrusted
+     * variable size - trusted
+     * variable p - trusted
+     */
+
+    /* GOOD */
+    if (len \ size - p) return EINVAL;
```
Comment (pbrezina) [inline src/contrib/coding-style.rst:429]: /

File: src/contrib/coding-style.rst [hunk 2]
```diff
@@ -1,12 +1,494 @@
...
+
+    /* BAD */
+    if ((p + len ) \ size) return EINVAL;
```
Comment (pbrezina) [inline src/contrib/coding-style.rst:432]: /

**Review by pbrezina — CHANGES_REQUESTED** [review]
Please reference it from `contents.rst`.

**Comment by pbrezina** [comment]
@justin-stephenson Can you please check the language?

File: src/contrib/coding-style.rst [hunk 3]
```diff
@@ -1,12 +1,494 @@
 Coding Style
+############
+
+Introduction
 ============
+This manual contains coding standards and guidelines for SSSD contributors.
+All new code should adhere to these standards. Those rules does not apply to
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:7]: s/Those rules does not/These rules do not/

File: src/contrib/coding-style.rst [hunk 4]
```diff
@@ -1,12 +1,494 @@
...
+existing legacy code. Please do not go back and apply formatting changes to
+the old code. It just confuses things and is generally a waste of time.
+
+Why we are using coding guidelines?
+-----------------------------------
+The Coding Guidelines are necessary to improve maintainability and readability
+of the code. It allows contributors from all over the world to stick to the one
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:14]: s/stick to the one/stick to one/

File: src/contrib/coding-style.rst [hunk 5]
```diff
@@ -1,12 +1,494 @@
...
+code convention making code colaboration much easier.
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:15]: s/code colaboration/collaboration/

File: src/contrib/coding-style.rst [hunk 6]
```diff
@@ -1,12 +1,494 @@
...
+
+Improved readability
+--------------------
+Code is not only for compilers, is mainly for peoples too.
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:19]: s/is mainly for peoples too/it is for people also/

File: src/contrib/coding-style.rst [hunk 7]
```diff
@@ -1,12 +1,494 @@
...
+Coding style and consistency means that if you go from one part of the code
+to another you don’t spend time having to re-adjust from one style to another.
+Blocks are consistent and readable and the flow of the code is apparent.
+Coding style adds value for people who have to read your code after you’ve been
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:23]: s/who have to read/reading/

File: src/contrib/coding-style.rst [hunk 8]
```diff
@@ -1,12 +1,494 @@
...
+hit by a `bus <https://en.wikipedia.org/wiki/Bus_factor>`_.
+Remember that.
+
+Easier maintanance
+------------------
+If you’re merging changes from a patch it’s much easier if everyone is using
+the same coding style. This isn’t the reality for a lot of our code,
+but we’re trying to get to the point where most of it uses the same style.
+The less time maintainer will spend on adjusting to the new coding style,
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:32]: s/maintainer/maintainers/

File: src/contrib/coding-style.rst [hunk 9]
```diff
@@ -1,12 +1,494 @@
...
+
+Code structure
+---------------
+- All source and header files should include a copy of the license.
+- For best readability line length should be 80 characters, in justified cases
+  maximum line length can be up to 120 characters.
+
+.. code-block:: c
+
+    /* Bellow situation is common and accepted */
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:57]: s/Bellow/Below/

File: src/contrib/coding-style.rst [hunk 10]
```diff
@@ -1,12 +1,494 @@
...
+    ret = friendly_function_with_meaningful_name(arg1,
+                                                 arg2,
+                                                 arg3)
+
+    /* Bellow situation is accepted but should be avoided */
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:62]: s/Bellow/Below/

File: src/contrib/coding-style.rst [hunk 11]
```diff
@@ -1,12 +1,494 @@
...
+    ret = this_is_some_function_with_very_long_name_from_external_library_and_cant_fit_80_characters(arg1,
+                                                                                                     arg2,
+                                                                                                     arg3)
+
+- When wrapping lines try to break it in common place:
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:67]: s/in common/in a common/

File: src/contrib/coding-style.rst [hunk 12]
```diff
@@ -1,12 +1,494 @@
...
+
+  - after a comma
+  - before an operator
+  - align the new line with the beginning of the expression at the same level in the previous line
+  - if all above fails, just indent 8 spaces
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:72]: s/just//

File: src/contrib/coding-style.rst [hunk 13]
```diff
@@ -1,12 +1,494 @@
...
+          arg3);
+
+Comments
+--------
+- C++ style line comments "// ......" should be avoided. Avoid useless comments
+  that do not add value to the code.
+
+.. code-block:: c
+
+    // I am bad multi line comment
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:92]: s/bad/a bad/

File: src/contrib/coding-style.rst [hunk 14]
```diff
@@ -1,12 +1,494 @@
...
+    // and should not look like this
+    // so please avoid doing something like me
+    void fun(int arg1, int arg2, int arg3); // I also should not be here
+
+    /*
+     * I am bad multi line comment
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:98]: s/bad/a bad/

File: src/contrib/coding-style.rst [hunk 15]
```diff
@@ -1,12 +1,494 @@
...
+        tevent_req_error(req, ret);
+    }
+
+    tevent_req_post(req, ev);
+    return req;
+
+Variables
+=========
+- Never use Hungarian notation when naming variables.
+- Use low case multi word underscore separated notation for naming variables.
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:263]: s/low/lower/

File: src/contrib/coding-style.rst [hunk 16]
```diff
@@ -1,12 +1,494 @@
...
+- Make variable name meaningful.
+- Always declare variables at the top of the function or block. If you find
+  yourself declaring many variables inside inner block or loop, consider
+  refactoring the block into helper function.
+- One declaration per line is preferred.
+- Initialize loval variables at declaration time when possible.
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:269]: s/loval/lval/ ?

File: src/contrib/coding-style.rst [hunk 17]
```diff
@@ -1,12 +1,494 @@
...
+     */
+
+    /* GOOD */
+    if (len \ size - p) return EINVAL;
+
+    /* BAD */
+    if ((p + len ) \ size) return EINVAL;
+
+
+- Always have default case in switch() statement
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:435]: s/statement/statements/

File: src/contrib/coding-style.rst [hunk 18]
```diff
@@ -1,12 +1,494 @@
...
+- Add comments if missing break is intentional
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:436]: s/missing/a missing/

**Comment by justin-stephenson** [comment]
Small grammatical changes requested, thank you.

File: src/contrib/coding-style.rst (hunk 1 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:429]: Fixed

File: src/contrib/coding-style.rst (hunk 2 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:432]: Fixed

File: src/contrib/coding-style.rst (hunk 16 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:269]: It mean to be "local variables" :) Fixed

File: src/contrib/coding-style.rst (hunk 18 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:436]: Fixed

File: src/contrib/coding-style.rst (hunk 17 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:435]: Fixed

File: src/contrib/coding-style.rst (hunk 16 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:269]: Fixed

File: src/contrib/coding-style.rst (hunk 15 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:263]: Fixed

File: src/contrib/coding-style.rst (hunk 14 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:98]: Fixed

File: src/contrib/coding-style.rst (hunk 13 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:92]: Fixed

File: src/contrib/coding-style.rst (hunk 12 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:72]: Fixed

File: src/contrib/coding-style.rst (hunk 11 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:67]: Fixed

File: src/contrib/coding-style.rst (hunk 10 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:62]: Fixed

File: src/contrib/coding-style.rst (hunk 9 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:57]: Fixed

File: src/contrib/coding-style.rst (hunk 8 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:32]: Fixed

File: src/contrib/coding-style.rst (hunk 7 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:23]: Fixed

File: src/contrib/coding-style.rst (hunk 6 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:19]: Fixed

File: src/contrib/coding-style.rst (hunk 5 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:15]: Fixed

File: src/contrib/coding-style.rst (hunk 4 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:14]: Fixed

File: src/contrib/coding-style.rst (hunk 3 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst:7]: Fixed

**Comment by elkoniu** [comment]
> Please reference it from `contents.rst`.

What do you mean by this? It is already linked via Contribute > Coding Style

**Comment by pbrezina** [comment]
> > Please reference it from `contents.rst`.
> 
> What do you mean by this? It is already linked via Contribute > Coding Style

Ah, sorry. I missed that it is already there from the initial commit.

File: src/contrib/coding-style.rst [hunk 19]
```diff
@@ -1,12 +1,502 @@
...
+    /* Bad message */
+    LOG_DEBUG("Error!!!");
+
+    /* Better message */
+    LOG_DEBUG("An error occured processing client request");
+
+    /* Great message */
+    LOG_DEBUG("An error occured processing client request: <ID>, result: <RET>");
+
+    /* Too great message */
```
Comment (justin-stephenson) [inline src/contrib/coding-style.rst:118]: Too long of a message?

**Comment by justin-stephenson** [comment]
Added one more comment but otherwise Ack from my side.

File: src/contrib/coding-style.rst (hunk 19 above)
Comment (elkoniu) [reply src/contrib/coding-style.rst]: Fixed

**Comment by elkoniu** [comment]
@justin-stephenson All requested changes has been applied. If you are OK with final result please set ACCEPT flag.

**Comment by justin-stephenson** [comment]
> @justin-stephenson All requested changes has been applied. If you are OK with final result please set ACCEPT flag.

Ack from me but I wasn't sure if Pavel had any other comments, that's why I haven't added `Accept` label

**Comment by pbrezina** [comment]
Ack. Thank you.

**Pull request merged**